--- a/renpy/module/ffmedia.c
+++ b/renpy/module/ffmedia.c
@@ -56,17 +56,26 @@
 
 /*******************************************************************************
  * SDL_RWops <-> AVIOContext
- * */
+ */
 
 static int rwops_read(void *opaque, uint8_t *buf, int buf_size) {
     SDL_RWops *rw = (SDL_RWops *) opaque;
 
     int rv = rw->read(rw, buf, 1, buf_size);
+
+	if (rv == 0) {
+		return AVERROR_EOF;
+	}
+
     return rv;
 
 }
 
+#if (LIBAVFORMAT_VERSION_MAJOR < 61)
 static int rwops_write(void *opaque, uint8_t *buf, int buf_size) {
+#else
+static int rwops_write(void *opaque, const uint8_t *buf, int buf_size) {
+#endif
     printf("Writing to an SDL_rwops is a really bad idea.\n");
     return -1;
 }
@@ -116,9 +125,14 @@
 
 static double current_time = 0;
 
+typedef struct PacketQueueEntry {
+	AVPacket *pkt;
+	struct PacketQueueEntry *next;
+} PacketQueueEntry;
+
 typedef struct PacketQueue {
-	AVPacketList *first;
-	AVPacketList *last;
+	PacketQueueEntry *first;
+	PacketQueueEntry *last;
 } PacketQueue;
 
 typedef struct FrameQueue {
@@ -200,13 +214,10 @@
 	AVCodecContext *video_context;
 	AVCodecContext *audio_context;
 
-	/* Queues of packets going to the audio and video
-	 * streams.
-	 */
+	/* Queues of packets going to the audio and video streams. */
 	PacketQueue video_packet_queue;
 	PacketQueue audio_packet_queue;
 
-
 	/* The total duration of the video. Only used for information purposes. */
 	double total_duration;
 
@@ -239,11 +250,6 @@
 	/* A frame that video is decoded into. */
 	AVFrame *video_decode_frame;
 
-	/* The video packet we're decoding, and the partial packet. */
-	AVPacket video_pkt;
-	AVPacket video_pkt_tmp;
-
-
 	/* Video Stuff ***********************************************************/
 
 	/* Software rescaling context. */
@@ -268,7 +274,6 @@
 	/* The offset between now and the time of the current frame, at least for video. */
 	double time_offset;
 
-
 } MediaState;
 
 static AVFrame *dequeue_frame(FrameQueue *fq);
@@ -310,8 +315,6 @@
 		av_frame_free(&ms->video_decode_frame);
 	}
 
-	av_packet_unref(&ms->video_pkt);
-
 	/* Destroy audio stuff. */
 	if (ms->swr) {
 		swr_free(&ms->swr);
@@ -404,6 +407,8 @@
     SDL_UnlockMutex(deallocate_mutex);
 }
 
+/* Frame queue ***************************************************************/
+
 static void enqueue_frame(FrameQueue *fq, AVFrame *frame) {
 	frame->opaque = NULL;
 
@@ -431,68 +436,109 @@
 }
 
 
+/* Packet queue **************************************************************/
+
 static void enqueue_packet(PacketQueue *pq, AVPacket *pkt) {
-	AVPacketList *pl = av_malloc(sizeof(AVPacketList));
-	if (pl == NULL)
-	{
+	PacketQueueEntry *pqe = av_malloc(sizeof(PacketQueueEntry));
+	if (pqe == NULL) {
+		av_packet_free(&pkt);
 		return;
 	}
 
-	av_init_packet(&pl->pkt);
-	av_packet_ref(&pl->pkt, pkt);
+	pqe->pkt = pkt;
+	pqe->next = NULL;
 
-	pl->next = NULL;
-
 	if (!pq->first) {
-		pq->first = pq->last = pl;
+		pq->first = pq->last = pqe;
 	} else {
-		pq->last->next = pl;
-		pq->last = pl;
+		pq->last->next = pqe;
+		pq->last = pqe;
 	}
 }
 
-static int dequeue_packet(PacketQueue *pq, AVPacket *pkt) {
-	if (! pq->first ) {
-		return 0;
+static AVPacket *first_packet(PacketQueue *pq) {
+	if (pq->first) {
+		return pq->first->pkt;
+	} else {
+		return NULL;
 	}
+}
 
-	AVPacketList *pl = pq->first;
+static void dequeue_packet(PacketQueue *pq) {
+	if (! pq->first) {
+		return;
+	}
 
-	av_packet_move_ref(pkt, &pl->pkt);
+	PacketQueueEntry *pqe = pq->first;
+	pq->first = pqe->next;
 
-	pq->first = pl->next;
-
 	if (!pq->first) {
 		pq->last = NULL;
 	}
 
-	av_free(pl);
-
-	return 1;
+	av_packet_free(&pqe->pkt);
+	av_free(pqe);
 }
+
 static int count_packet_queue(PacketQueue *pq) {
-       AVPacketList *pl = pq->first;
+    PacketQueueEntry *pqe = pq->first;
 
-       int rv = 0;
+	int rv = 0;
 
-       while (pl) {
-               rv += 1;
-               pl = pl->next;
-       }
+	while (pqe) {
+		rv += 1;
+		pqe = pqe->next;
+	}
 
-       return rv;
+	return rv;
 }
 
 static void free_packet_queue(PacketQueue *pq) {
-	AVPacket scratch;
+	while(first_packet(pq)) {
+		dequeue_packet(pq);
+	}
+}
 
-	av_init_packet(&scratch);
 
-	while (dequeue_packet(pq, &scratch)) {
-		av_packet_unref(&scratch);
+/**
+ * Reads a packet from one of the queues, filling the other queue if
+ * necessary. Returns the packet, or NULL if end of file has been reached.
+ */
+static AVPacket *read_packet(MediaState *ms, PacketQueue *pq) {
+
+	AVPacket *pkt;
+	AVPacket *rv;
+
+	while (1) {
+
+		rv = first_packet(pq);
+		if (rv) {
+			return rv;
+		}
+
+		pkt = av_packet_alloc();
+
+		if (!pkt) {
+			return NULL;
+		}
+
+		if (av_read_frame(ms->ctx, pkt)) {
+			return NULL;
+		}
+
+		if (pkt->stream_index == ms->video_stream && ! ms->video_finished) {
+			enqueue_packet(&ms->video_packet_queue, pkt);
+		} else if (pkt->stream_index == ms->audio_stream && ! ms->audio_finished) {
+			enqueue_packet(&ms->audio_packet_queue, pkt);
+		} else {
+			av_packet_free(&pkt);
+		}
 	}
 }
 
+
+/* Surface queue *************************************************************/
+
 static void enqueue_surface(SurfaceQueueEntry **queue, SurfaceQueueEntry *sqe) {
 	while (*queue) {
 		queue = &(*queue)->next;
@@ -532,38 +578,9 @@
 }
 #endif
 
+/* Find decoder context ******************************************************/
 
-/**
- * Reads a packet from one of the queues, filling the other queue if
- * necessary.
- */
-static int read_packet(MediaState *ms, PacketQueue *pq, AVPacket *pkt) {
-	AVPacket scratch;
 
-	av_init_packet(&scratch);
-
-	while (1) {
-		if (dequeue_packet(pq, pkt)) {
-			return 1;
-		}
-
-		if (av_read_frame(ms->ctx, &scratch)) {
-			pkt->data = NULL;
-			pkt->size = 0;
-			return 0;
-		}
-
-		if (scratch.stream_index == ms->video_stream && ! ms->video_finished) {
-			enqueue_packet(&ms->video_packet_queue, &scratch);
-		} else if (scratch.stream_index == ms->audio_stream && ! ms->audio_finished) {
-			enqueue_packet(&ms->audio_packet_queue, &scratch);
-		}
-
-		av_packet_unref(&scratch);
-	}
-}
-
-
 static AVCodecContext *find_context(AVFormatContext *ctx, int index) {
 
     AVDictionary *opts = NULL;
@@ -572,7 +589,7 @@
 		return NULL;
 	}
 
-	AVCodec *codec = NULL;
+	const AVCodec *codec = NULL;
 	AVCodecContext *codec_ctx = NULL;
 
 	codec_ctx = avcodec_alloc_context3(NULL);
@@ -610,65 +627,14 @@
 
 	avcodec_free_context(&codec_ctx);
 	return NULL;
-}
-
-
-/**
- * Given a packet, decodes a frame if possible. This is intended to be a drop-in replacement
- * for the now deprecated avcodec_decode_audio4/video2 APIs.
- *
- * \param[in]   context     The context the decoding is done in.
- * \param[out]  frame       A frame that is updated with the decoded data.
- * \param[out]  got_frame   Set to 1 if a frame was decoded, 0 if not.
- * \param[in]   pkt         The packet data to present.
- *
- * Returns pkt->size if the packet was consumed, 0 if not, or < 0 on error (including
- * end of file.)
- */
-static int decode_common(AVCodecContext *context, AVFrame *frame, int *got_frame, AVPacket *pkt) {
-
-    int ret;
-    int rv = 0;
-
-    if (pkt) {
-        ret = avcodec_send_packet(context, pkt);
-
-        if (ret >= 0) {
-            rv = pkt->size;
-        } else if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
-            rv = 0;
-        } else {
-           return ret;
-        }
-    }
-
-    ret = avcodec_receive_frame(context, frame);
-
-    if (ret >= 0) {
-        *got_frame = 1;
-    } else if (ret == AVERROR(EAGAIN)) {
-        *got_frame = 0;
-    } else if (ret == AVERROR_EOF) {
-        *got_frame = 0;
-        if (!pkt || pkt->size == 0) {
-            return ret;
-        }
-    } else {
-        *got_frame = 0;
-        return ret;
-    }
-
-    return rv;
 }
 
 
-/**
- * Decodes audio. Returns 0 if no audio was decoded, or 1 if some audio was
- * decoded.
- */
+/* Audio decoding *************************************************************/
+
 static void decode_audio(MediaState *ms) {
-	AVPacket pkt;
-	AVPacket pkt_temp;
+	int ret;
+	AVPacket *pkt;
 	AVFrame *converted_frame;
 
 	if (!ms->audio_context) {
@@ -685,8 +651,6 @@
 		return;
 	}
 
-	av_init_packet(&pkt);
-
 	double timebase = av_q2d(ms->ctx->streams[ms->audio_stream]->time_base);
 
 	if (ms->audio_queue_target_samples < audio_target_samples) {
@@ -695,37 +659,33 @@
 
 	while (ms->audio_queue_samples < ms->audio_queue_target_samples) {
 
-		read_packet(ms, &ms->audio_packet_queue, &pkt);
+		/** Read a packet, and send it to the decoder. */
+		pkt = read_packet(ms, &ms->audio_packet_queue);
+		ret = avcodec_send_packet(ms->audio_context, pkt);
 
-		pkt_temp = pkt;
+		if (ret == 0) {
+			dequeue_packet(&ms->audio_packet_queue);
+		} else if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
+			// pass
+		} else {
+			ms->audio_finished = 1;
+			return;
+		}
 
-		do {
-			int got_frame;
-			int read_size = decode_common(ms->audio_context, ms->audio_decode_frame, &got_frame, &pkt_temp);
+		while (1) {
 
-			if (read_size < 0) {
+			ret = avcodec_receive_frame(ms->audio_context, ms->audio_decode_frame);
 
-				if (pkt.data) {
-					av_packet_unref(&pkt);
-				}
+			// More input is needed.
+			if (ret == AVERROR(EAGAIN)) {
+				break;
+			}
 
+			if (ret < 0) {
 				ms->audio_finished = 1;
 				return;
 			}
 
-			pkt_temp.data += read_size;
-			pkt_temp.size -= read_size;
-
-			if (!got_frame) {
-				if (pkt.data == NULL) {
-					ms->audio_finished = 1;
-					av_packet_unref(&pkt);
-					return;
-				}
-
-				break;
-			}
-
             converted_frame = av_frame_alloc();
 
 			if (converted_frame == NULL) {
@@ -734,9 +694,14 @@
 			}
 
             converted_frame->sample_rate = audio_sample_rate;
+#if (LIBAVUTIL_VERSION_MAJOR < 59)
             converted_frame->channel_layout = AV_CH_LAYOUT_STEREO;
+#else
+            converted_frame->ch_layout = (AVChannelLayout) AV_CHANNEL_LAYOUT_STEREO;
+#endif
             converted_frame->format = AV_SAMPLE_FMT_S16;
 
+#if (LIBAVUTIL_VERSION_MAJOR < 59)
 			if (!ms->audio_decode_frame->channel_layout) {
 				ms->audio_decode_frame->channel_layout = av_get_default_channel_layout(ms->audio_decode_frame->channels);
 
@@ -755,7 +720,27 @@
 				    swr_set_matrix(ms->swr, stereo_matrix, 1);
 				}
 			}
+#else
+			if (ms->audio_decode_frame->ch_layout.order == AV_CHANNEL_ORDER_UNSPEC) {
+				av_channel_layout_default(&ms->audio_decode_frame->ch_layout, ms->audio_decode_frame->ch_layout.nb_channels);
 
+				if (audio_equal_mono && (ms->audio_decode_frame->ch_layout.nb_channels == 1)) {
+				    swr_alloc_set_opts2(
+                        &ms->swr,
+                        &converted_frame->ch_layout,
+                        converted_frame->format,
+                        converted_frame->sample_rate,
+                        &ms->audio_decode_frame->ch_layout,
+                        ms->audio_decode_frame->format,
+                        ms->audio_decode_frame->sample_rate,
+                        0,
+                        NULL);
+
+				    swr_set_matrix(ms->swr, stereo_matrix, 1);
+				}
+			}
+#endif
+
 			if(swr_convert_frame(ms->swr, converted_frame, ms->audio_decode_frame)) {
 				av_frame_free(&converted_frame);
 				continue;
@@ -785,18 +770,17 @@
 			}
 
 			SDL_UnlockMutex(ms->lock);
-
-		} while (pkt_temp.size);
-
-		if (pkt.data) {
-			av_packet_unref(&pkt);
 		}
+
 	}
 
 	return;
 
 }
 
+
+/* Video decoding *************************************************************/
+
 static enum AVPixelFormat get_pixel_format(SDL_Surface *surf) {
     uint32_t pixel;
     uint8_t *bytes = (uint8_t *) &pixel;
@@ -819,37 +803,37 @@
 }
 
 
-
 static SurfaceQueueEntry *decode_video_frame(MediaState *ms) {
+	int ret;
 
 	while (1) {
 
-		if (! ms->video_pkt_tmp.size) {
-			av_packet_unref(&ms->video_pkt);
-			read_packet(ms, &ms->video_packet_queue, &ms->video_pkt);
-			ms->video_pkt_tmp = ms->video_pkt;
-		}
+		AVPacket *pkt = read_packet(ms, &ms->video_packet_queue);
+		ret = avcodec_send_packet(ms->video_context, pkt);
 
-		int got_frame = 0;
-		int read_size = decode_common(ms->video_context, ms->video_decode_frame, &got_frame, &ms->video_pkt_tmp);
 
-		if (read_size < 0) {
+		if (ret == 0) {
+			dequeue_packet(&ms->video_packet_queue);
+		} else if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
+			// pass
+		} else {
 			ms->video_finished = 1;
 			return NULL;
 		}
 
-		ms->video_pkt_tmp.data += read_size;
-		ms->video_pkt_tmp.size -= read_size;
+		ret = avcodec_receive_frame(ms->video_context, ms->video_decode_frame);
 
-		if (got_frame) {
-			break;
+		// More input is needed.
+		if (ret == AVERROR(EAGAIN)) {
+			continue;
 		}
 
-		if (!got_frame && !ms->video_pkt.size) {
+		if (ret < 0) {
 			ms->video_finished = 1;
 			return NULL;
 		}
 
+		break;
 	}
 
 	double pts = ms->video_decode_frame->best_effort_timestamp * av_q2d(ms->ctx->streams[ms->video_stream]->time_base);
@@ -1192,12 +1176,14 @@
 	if (ms->swr == NULL) {
 		goto finish;
 	}
-
-	av_init_packet(&ms->video_pkt);
 
 	// Compute the number of samples we need to play back.
 	if (ms->audio_duration < 0) {
+#if (LIBAVFORMAT_VERSION_MAJOR < 62)
 		if (av_fmt_ctx_get_duration_estimation_method(ctx) != AVFMT_DURATION_FROM_BITRATE) {
+#else
+		if (ctx->duration_estimation_method != AVFMT_DURATION_FROM_BITRATE) {
+#endif
 
 			long long duration = ((long long) ctx->duration) * audio_sample_rate;
 			ms->audio_duration = (unsigned int) (duration /  AV_TIME_BASE);
@@ -1354,12 +1340,14 @@
 	if (ms->swr == NULL) {
 		media_read_sync_finish(ms);
 	}
-
-	av_init_packet(&ms->video_pkt);
 
 	// Compute the number of samples we need to play back.
 	if (ms->audio_duration < 0) {
+#if (LIBAVFORMAT_VERSION_MAJOR < 62)
 		if (av_fmt_ctx_get_duration_estimation_method(ctx) != AVFMT_DURATION_FROM_BITRATE) {
+#else
+		if (ctx->duration_estimation_method != AVFMT_DURATION_FROM_BITRATE) {
+#endif
 
 			long long duration = ((long long) ctx->duration) * audio_sample_rate;
 			ms->audio_duration = (unsigned int) (duration /  AV_TIME_BASE);
